server/src/main/java/org/elasticsearch/transport/Lz4TransportDecompressor.java

```diff
@@ -0,0 +1,355 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+/*
+ * Copyright 2014 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License, version
+ * 2.0 (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.elasticsearch.transport;
+
+import net.jpountz.lz4.LZ4Exception;
+import net.jpountz.lz4.LZ4Factory;
+import net.jpountz.lz4.LZ4FastDecompressor;
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.bytes.BytesArray;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.bytes.ReleasableBytesReference;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.recycler.Recycler;
+import org.elasticsearch.common.util.PageCacheRecycler;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Locale;
+import java.util.zip.Checksum;
+
+/**
+ * This file is forked from the https://netty.io project. In particular it forks the follow file
+ * io.netty.handler.codec.compression.Lz4FrameDecoder.
+ *
+ * It modifies the original netty code to operate on byte arrays opposed to ByteBufs.
+ * Additionally, it integrates the decompression code to work in the Elasticsearch transport
+ * pipeline, Finally, it replaces the custom Netty decoder exceptions.
+ */
```

Comment:
An elaborate comment explaining why this is necessary and where its compressing counterpart would be good to add.